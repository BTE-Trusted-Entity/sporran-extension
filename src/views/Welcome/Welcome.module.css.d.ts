export const __esModule: true;
export const container: string;
export const heading: string;
export const info: string;
export const termsLine: string;
export const agreeLabel: string;
export const agree: string;
export const terms: string;
export const create: string;
export const importPhrase: string;
