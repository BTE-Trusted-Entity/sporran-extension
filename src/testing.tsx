export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
