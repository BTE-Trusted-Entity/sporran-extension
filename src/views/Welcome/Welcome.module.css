.container {
  composes: container from '../../components/View/View.module.css';
  justify-content: center;
  height: 100%;
  padding-bottom: 5rem;
  box-sizing: border-box;
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.info {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin-top: 0.375rem;
  margin-bottom: 0;
  white-space: pre-line;
  line-height: 1.25rem;
  text-align: center;
}

.termsLine {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin-top: 1.25rem;
}

.agreeLabel {
  cursor: pointer;
}

.agree {
  composes: checkbox from '../../components/Checkbox/Checkbox.module.css';
}

.terms {
  composes: subline from '../../components/Typography/Typography.module.css';
}

.terms:not(:hover) {
  text-decoration: none;
}

.create {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.importPhrase {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
  margin-top: 1.25rem;
}
