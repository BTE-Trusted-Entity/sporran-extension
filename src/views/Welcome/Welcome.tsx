import {
  ChangeEvent,
  Fragment,
  MouseEvent,
  useCallback,
  useState,
} from 'react';
import { browser } from 'webextension-polyfill-ts';
import { Link, Redirect } from 'react-router-dom';

import * as styles from './Welcome.module.css';

import {
  useIdentities,
  useCurrentIdentity,
} from '../../utilities/identities/identities';
import { YouHaveIdentities } from '../../components/YouHaveIdentities/YouHaveIdentities';
import { LinkBack } from '../../components/LinkBack/LinkBack';
import { generatePath, paths } from '../paths';

interface Props {
  again?: boolean;
}

export function Welcome({ again = false }: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const [enabled, setEnabled] = useState(false);
  const handleTermsClick = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      setEnabled(event.target.checked);
    },
    [],
  );

  const handleLinkClick = useCallback(
    (event: MouseEvent) => {
      if (!enabled) {
        event.preventDefault();
      }
    },
    [enabled],
  );

  const identities = useIdentities();
  const current = useCurrentIdentity();

  if (!identities.data) {
    return null; // storage data pending
  }

  const identitiesNumber = Object.values(identities.data).length;
  const hasIdentities = identitiesNumber > 0;

  if (current && hasIdentities && !again) {
    return (
      <Redirect
        to={generatePath(
          paths.identity.overview,
          identities.data[current] as { address: string },
        )}
      />
    );
  }

  return (
    <div className={styles.container}>
      <h1 className={styles.heading}>
        {hasIdentities ? t('view_Welcome_again') : t('view_Welcome_heading')}
      </h1>

      <h3 className={styles.info}>
        {hasIdentities ? (
          <Fragment>
            <YouHaveIdentities />
            {t('view_Welcome_addAnother')}
          </Fragment>
        ) : (
          t('view_Welcome_noIdentities')
        )}
      </h3>

      <p className={styles.termsLine}>
        <label className={styles.agreeLabel}>
          <input
            className={styles.agree}
            type="checkbox"
            onChange={handleTermsClick}
            checked={enabled}
          />
          <span />
          {t('view_Welcome_agree')}
        </label>
        <a
          className={styles.terms}
          href="https://www.sporran.org/terms"
          target="_blank"
          rel="noreferrer"
        >
          {t('view_Welcome_terms')}
        </a>
      </p>

      <Link
        to={paths.identity.create.start}
        className={styles.create}
        onClick={handleLinkClick}
        aria-disabled={!enabled}
      >
        {t('view_Welcome_create')}
      </Link>

      <Link
        to={paths.identity.import.start}
        className={styles.importPhrase}
        onClick={handleLinkClick}
        aria-disabled={!enabled}
      >
        {t('view_Welcome_import')}
      </Link>

      {again && <LinkBack />}
    </div>
  );
}
