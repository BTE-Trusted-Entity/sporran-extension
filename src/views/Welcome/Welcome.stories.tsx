import { Meta } from '@storybook/react';

import { IdentitiesProviderMock } from '../../utilities/identities/IdentitiesProvider.mock';

import { Welcome } from './Welcome';

export default {
  title: 'Views/Welcome',
  component: Welcome,
} as Meta;

export function NoIdentities(): JSX.Element {
  return (
    <IdentitiesProviderMock identities={{}}>
      <Welcome />
    </IdentitiesProviderMock>
  );
}

export function HasIdentities(): JSX.Element {
  return <Welcome again />;
}
