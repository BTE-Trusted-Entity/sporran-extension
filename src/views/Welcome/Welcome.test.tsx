import userEvent from '@testing-library/user-event';

import { identitiesMock, render, screen } from '../../testing/testing';
import {
  useCurrentIdentity,
  useIdentities,
} from '../../utilities/identities/identities';

import { Welcome } from './Welcome';

jest.mock('../../utilities/identities/identities');
jest
  .mocked(useCurrentIdentity)
  .mockReturnValue('4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1');

describe('Welcome', () => {
  it('should match the snapshot without identities', () => {
    jest.mocked(useIdentities).mockReturnValue({ data: {} });
    const { container } = render(<Welcome />);
    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot with an identity', () => {
    jest.mocked(useIdentities).mockReturnValue({ data: identitiesMock });
    const { container } = render(<Welcome again />);
    expect(container).toMatchSnapshot();
  });

  it('should enable links after accepting T&C', async () => {
    jest.mocked(useIdentities).mockReturnValue({ data: identitiesMock });
    render(<Welcome again />);

    const createLink = await screen.findByText(/Create/);
    const importLink = await screen.findByText(/Import/);

    expect(createLink).toHaveAttribute('aria-disabled', 'true');
    expect(importLink).toHaveAttribute('aria-disabled', 'true');

    await userEvent.click(await screen.findByLabelText(/agree to the/));

    expect(createLink).not.toHaveAttribute('aria-disabled', 'true');
    expect(importLink).not.toHaveAttribute('aria-disabled', 'true');
  });
});
