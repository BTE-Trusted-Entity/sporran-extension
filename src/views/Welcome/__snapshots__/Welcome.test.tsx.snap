// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome should match the snapshot with an identity 1`] = `
<div>
  <div
    class="container"
  >
    <h1
      class="heading"
    >
      Add Identity
    </h1>
    <h3
      class="info"
    >
      You have 3 Identities
      . Add another one.
    </h3>
    <p
      class="termsLine"
    >
      <label
        class="agreeLabel"
      >
        <input
          class="agree"
          type="checkbox"
        />
        <span />
         I have read and agree to the 
      </label>
      <a
        class="terms"
        href="https://www.sporran.org/terms"
        rel="noreferrer"
        target="_blank"
      >
        Terms & Conditions
      </a>
    </p>
    <a
      aria-disabled="true"
      class="create"
      href="/identity/create"
    >
      Create Identity
    </a>
    <a
      aria-disabled="true"
      class="importPhrase"
      href="/identity/import"
    >
      Import an Identity from pre-existing phrase
    </a>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </div>
</div>
`;

exports[`Welcome should match the snapshot without identities 1`] = `
<div>
  <div
    class="container"
  >
    <h1
      class="heading"
    >
      Welcome to your Sporran
    </h1>
    <h3
      class="info"
    >
      You currently donâ€™t have any Identities.
Add your first Identity to get started.
    </h3>
    <p
      class="termsLine"
    >
      <label
        class="agreeLabel"
      >
        <input
          class="agree"
          type="checkbox"
        />
        <span />
         I have read and agree to the 
      </label>
      <a
        class="terms"
        href="https://www.sporran.org/terms"
        rel="noreferrer"
        target="_blank"
      >
        Terms & Conditions
      </a>
    </p>
    <a
      aria-disabled="true"
      class="create"
      href="/identity/create"
    >
      Create Identity
    </a>
    <a
      aria-disabled="true"
      class="importPhrase"
      href="/identity/import"
    >
      Import an Identity from pre-existing phrase
    </a>
  </div>
</div>
`;
